Lib/inspect.py

```diff
@@ -3355,11 +3397,135 @@ def format(self, *, max_width=None):
+        try:
+            return self._return_annotation
+        except AttributeError:
+            pass
+        return_annotations = []
+        for s in self._signatures:
+            ann = s.return_annotation
+            if ann != _empty:
+                if ann is None:
+                    ann = type(ann)
+                elif isinstance(ann, str):
+                    from typing import ForwardRef
+                    ann = ForwardRef(ann)
+                return_annotations.append(ann)
+        self._return_annotation = (functools.reduce(or_, return_annotations)
+                                   if return_annotations else _empty)
+        return self._return_annotation
+
+    def replace(self):
+        raise NotImplementedError
+
+    __replace__ = replace
+
+    def __iter__(self):
+        return iter(self._signatures)
+
+    def __hash__(self):
+        if len(self._signatures) == 1:
+            return hash(self._signatures[0])
+        return hash(self._signatures)
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if isinstance(other, MultiSignature):
+            return self._signatures == other._signatures
+        if len(self._signatures) == 1 and isinstance(other, Signature):
+            return self._signatures[0] == other
+        return NotImplemented
+
+    def bind(self, /, *args, **kwargs):
+        last_exc = None
+        for s in self._signatures:
+            try:
+                return s.bind(*args, **kwargs)
+            except TypeError as e:
+                last_exc = e
+        raise last_exc
+
+    def bind_partial(self, /, *args, **kwargs):
+        last_exc = None
+        bas = []
+        for s in self._signatures:
+            try:
+                bas.append(s.bind_partial(*args, **kwargs))
+            except TypeError as e:
+                last_exc = e
+        if not bas:
+            raise last_exc
+        if len(bas) == 1:
+            return bas[0]
+        arguments = bas[0].arguments
+        for ba in bas:
+            if ba.arguments != arguments:
+                raise TypeError('ambiguous binding')
+        sig = (self if len(bas) == len(self._signatures) else
+               self.__class__([ba.signature for ba in bas]))
+        return BoundArguments(sig, arguments)
+
+    def __reduce__(self):
+        return type(self), (self._signatures,)
+
+    def __repr__(self):
+        return '<%s %s>' % (self.__class__.__name__,
+                            '|'.join(map(str, self._signatures)))
+
+    def __str__(self):
+        return '\n'.join(map(str, self._signatures))
+
+    def format(self, *, max_width=None):
+        return '\n'.join(sig.format(max_width=max_width)
+                         for sig in self._signatures)
+
+
 def signature(obj, *, follow_wrapped=True, globals=None, locals=None, eval_str=False):
     """Get a signature object for the passed callable."""
     return Signature.from_callable(obj, follow_wrapped=follow_wrapped,
```

Comment:
Given that the original Signature.from_callable is not changed and does the exact same call to `_from_callable`, that means both `signature` and `Signature.from_callable` may return a MultiSignature object, right ?

Strictly speaking this is not a violation of backwards compatibility as MultiSignature is a subclass of Signature, but I find it weird and it may break some strict-type behaviors. Is there a way to have those only return pure Signature objects, without breaking everything this update brings to the table ?

Reply:
Yes, it is. The idea was that `inspect.signature()` will start to support previously not supported cases and return a MultiSignature which should be compatible with Signature. But now I am no longer sure. They are not so well compatible as I expected, and in most cases I needed to rewrite the code to explicitly handle MultiSignature. This is why I added `inspect.signatures()`. Now the code is in the intermediate state -- either MultiSignature will be made more compatible with Signature and `inspect.signatures()` may disappear, or they will became completely different things.